compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -1328,4 +1332,72 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
 
         Ok(ImplSourceConstDestructData { nested })
     }
+
+    /// Generates obligations for a lazy candidate
+    /// Where the obligations generated are all possible values for the type of a
+    /// `ConstKind::Unevaluated(..)`.
+    /// In the future, it would be nice to extend this to inductive proofs.
+    fn confirm_lazy_candidate(
+        &mut self,
+        obligation: &TraitObligation<'tcx>,
+        candidate: ty::PolyTraitPredicate<'tcx>,
+    ) -> Vec<PredicateObligation<'tcx>> {
+        let mut obligations = vec![];
+
+        let query = obligation.predicate.skip_binder().trait_ref.self_ty();
+        let ty::Adt(_adt_def, adt_substs) = query.kind() else {
+            unreachable!();
+        };
+        // Find one adt subst at a time, then will be handled recursively.
+        let const_to_replace = adt_substs
+            .consts()
+            .find(|ct| ct.ty().is_bool() && matches!(ct.kind(), ty::ConstKind::Unevaluated(..)))
+            .unwrap();
+
+        /// Folder for replacing specific const values in `substs`.
+        struct Folder<'tcx> {
+            tcx: TyCtxt<'tcx>,
+            replace: ty::Const<'tcx>,
+            with: ty::Const<'tcx>,
+        }
+
+        impl<'tcx> TypeFolder<'tcx> for Folder<'tcx> {
+            fn tcx<'a>(&'a self) -> TyCtxt<'tcx> {
+                self.tcx
+            }
+            fn fold_const(&mut self, c: ty::Const<'tcx>) -> ty::Const<'tcx> {
+                if c == self.replace { self.with } else { c }
+            }
+        }
+
+        let tcx = self.tcx();
+        for v in [true, false] {
```

Comment:
I've been giving this scheme some thought. Instead of generating all possible variants and proving them, maybe we could let the trait solver handle such "questions". Basically you are asking the trait solver whether

* `Type<X>: Trait` can be satisfied (where `X` is a constant)
* which is basically `for<const X: bool> Type<X>: Trait`
* which gets turned into something like `Type<PlaceholderConst>: Trait`
* which we could then treat differently from `Type<ParamConst>` or `Type<true>` or `Type<false>`

When we try to prove that, we could add some special handling for that case that essentially makes a query to list all possible impls for `Type<_>` even if normally it would avoid trying to match inference vars against possible impls. Once we got the list of all impls that match (for bools there should be either one (a generic one) or two (one `true` and one `false`)). I'm not sure this works out, but we should probably explore it before going with a feature that just lists all values. Especially because the listing would get annoying for enums with values

Reply:
I think this makes sense, as extending this to enums is also a pain. There are also cases which are inductive such as `0usize`, and `N.saturating_sub(1)`. Where would that kind of impl go?

As a counterpoint, this PR was meant to be a small part of this larger feature, but I think most use cases would be for bools and not enums with many variants.

Reply:
> I think most use cases would be for bools and not enums with many variants.

I'm not worried about enums with many variants. I was mainly thinking about an impl for `None` and one for `Some(T)`.

I get that this is supposed to be a minimal impl, but I don't know if we should land a minimal impl if we are sure it's going to get removed entirely and replaced by a rewrite (fwiw: I don't think we are sure yet, but I also don't know if anyone'll have the capacity to work on my explorative alternative proposed above).

Reply:
I can definitely make something which can have more cases added, but I am not entirely sure how to add something like that without explicit checks for certain traits. I've mostly been able to follow the trait selection documentation, since it mostly exists for candidate assembly, but beyond that point there isn't much.

Would it require building or refactoring a large part of the existing obligation system?

Reply:
i don't believe so. but i also don't know how to implement it exactly. this would be more exploratory work.

we could do a sync session on eploring the existing implementation and coming up with more concrete ideas.